### ISSUE TITLE
Black cursor in neovim with dark colorscheme

### ISSUE THREAD

**Issue by alexryndin** [alacritty/alacritty#4264]
For bug reports, the following information can help speed up the process. Please describe the bug that you have found and what you would expect to happen instead.

### System

OS: Arch Linux
Version: alacritty 0.5.0
Sway WM, Wayland.
Alacritty Config: https://github.com/alexryndin/dotfiles/blob/7cc014348e758edc56b6e7c4e37213ec89a61171/.config/alacritty/alacritty.yml

I prefer white and black colorscheme in alacritty, but I use dark scheme (gruvbox) in vim. It looks good outside vim:
![grim-27 09 2020-180138](https://user-images.githubusercontent.com/8342078/94368204-98743800-00eb-11eb-86f8-d6fd02fd1be0.png)
But cursor color fails to match vim's colorscheme (cursor is on line 2):
![grim-27 09 2020-180400](https://user-images.githubusercontent.com/8342078/94368273-cf4a4e00-00eb-11eb-8a46-e60826bbbc9f.png)
Compare to gnome-terminal:
![grim-27 09 2020-180511](https://user-images.githubusercontent.com/8342078/94368293-f86ade80-00eb-11eb-9739-f7d01c5e21bf.png)

**Comment by chrisduerr** [comment]
Then change your cursor color? What exactly are you looking for?

**Comment by alexryndin** [comment]
@chrisduerr I just want cursor to match background. If I go with your advice, I get following:
![grim-27 09 2020-182000](https://user-images.githubusercontent.com/8342078/94368696-3f59d380-00ee-11eb-8e54-43789bbda84c.png)
![grim-27 09 2020-182024](https://user-images.githubusercontent.com/8342078/94368700-47b20e80-00ee-11eb-8cf1-282d538e3b14.png)
"can't see" cursor outside vim, and very white in vim
Again, gnome-terminal is ok in both modes.

**Comment by kchibisov** [comment]
Then you want inverted cursor? Apply [this](https://github.com/alacritty/alacritty/blob/e9c0034f4d3ee003149fe5454942bea7ff3d98be/alacritty.yml#L194) configuration or comment out cursor section in colors

**Comment by alexryndin** [comment]
@kchibisov 
![grim-27 09 2020-184327](https://user-images.githubusercontent.com/8342078/94369255-53530480-00f1-11eb-95da-d3275446b3ac.png)

**Comment by kchibisov** [comment]
Is cursor always black like that, even when you cover colored text? I bet neovim changes the cursor color to some obscure value, could you try different color scheme?

**Comment by alexryndin** [comment]
@kchibisov I've updated the issue with some explanation.
I've got the same behavior in both vim and neovim.

**Comment by alexryndin** [comment]
@kchibisov colorscheme darkblue
![grim-27 09 2020-191453](https://user-images.githubusercontent.com/8342078/94369994-b6469a80-00f5-11eb-911d-b1051dad88df.png)

**Comment by alexryndin** [comment]
I also found out that cursor is black until at least one symbol appears in line. Consider:
1. Cursor is black. Important thing here is that cursor color is actually black outside vim. Maybe alacritty caches cursor color or something?
![grim-27 09 2020-201212](https://user-images.githubusercontent.com/8342078/94371257-d5492a80-00fd-11eb-957b-0b8bdcd3a9e0.png)
2. Insert some symbols. Cursor is white now
![grim-27 09 2020-201223](https://user-images.githubusercontent.com/8342078/94371269-e2661980-00fd-11eb-875d-bbdc64dccca6.png)
3. Delete all symbols. Cursor will **always be white** on this line.
![grim-27 09 2020-201231](https://user-images.githubusercontent.com/8342078/94371311-3ffa6600-00fe-11eb-91fe-3d11ee842947.png)
4. Try next line -- cursor is black.
![grim-27 09 2020-201617](https://user-images.githubusercontent.com/8342078/94371331-746e2200-00fe-11eb-96b7-82696f52c64e.png)

**Comment by kchibisov** [comment]
Try `TERM=xterm-256color neovim`

**Comment by alexryndin** [comment]
@kchibisov result:
![grim-27 09 2020-202324](https://user-images.githubusercontent.com/8342078/94371456-6a98ee80-00ff-11eb-8cb6-e66253f8de1a.png)
I think alacritty somehow caches actual cursor color, which is black. Please, consider my previous comment.

**Comment by kchibisov** [comment]
>I think alacritty somehow caches actual cursor color, which is black. Please, consider my previous comment.

No, it doesn't do that. Now try different colorscheme with `TERM=xterm-256color`. I was doing `nvim -u NONE`, `:set termguicolor`, `:colorscheme elfrod`. If you try without `TERM=xterm-256color` the cursor will be different

**Comment by alexryndin** [comment]
@kchibisov xterm-256color is my default $TERM, now I try `TERM=xterm-256color nvim`, black cursor, as expected:
![grim-27 09 2020-203246](https://user-images.githubusercontent.com/8342078/94371640-a4b6c000-0100-11eb-9623-c6172c3166da.png)
Then I set
 ```
:set termguicolors
:colorscheme elflord
```
And get:
![grim-27 09 2020-203123](https://user-images.githubusercontent.com/8342078/94371711-2c043380-0101-11eb-9fd8-fc4200eefff0.png)
black cursor!

**Comment by kchibisov** [comment]
The thing is that I can't repro that way, for me it's not black, if we use `nvim -u NONE`.

**Comment by alexryndin** [comment]
@kchibisov I used `-u NONE` in my previous test, sorry, I missed that.

**Comment by alexryndin** [comment]
@kchibisov Also please take into account that I __can__ get proper cursor color if I insert some symbols (and then delete them). See one of my previous comments:
![grim-27 09 2020-204746](https://user-images.githubusercontent.com/8342078/94371924-ba2ce980-0102-11eb-98f7-bc8cbb08f167.png)

**Comment by alexryndin** [comment]
@kchibisov when you try to reproduce the bug, what is your cursor color outside vim?

**Comment by alexryndin** [comment]
@kchibisov I've updated the issue with steps to reproduce. Could you try?

**Comment by chrisduerr** [comment]
I'm pretty sure @kchibisov was able to reproduce, it's just about finding a minimal repro at the moment that doesn't cause the same issue in other terminal emulators.

**Comment by alexryndin** [comment]
@chrisduerr yesterday I've updated the issue with a minimal repro, which doesn't cause the same problem at least in gnome-terminal. Please, could you go through the steps I've described and see if you get same behavior?

**Comment by chrisduerr** [comment]
> yesterday I've updated the issue with a minimal repro, which doesn't cause the same problem at least in gnome-terminal. Please, could you go through the steps I've described and see if you get same behavior?

That is not minimal. I have no problem with finding that minimal repro myself, but please be patient.

**Comment by alexryndin** [comment]
@chrisduerr I just thought you missed my update.
> That is not minimal
> I have no problem with finding that minimal repro myself.
I'm sorry, but I just don't get it. What is a minimal repro? Isn't a three steps with a 5loc config a minimal repro. I just want to provide you with a repro you would find minimal.

**Comment by chrisduerr** [comment]
A minimal repro would provide just the escape sequences necessary to create this problem. @kchibisov was already looking into this using the ref test, so that should hopefully be sufficient.

**Comment by kchibisov** [comment]
@chrisduerr it seems like that's the reproducer. It's pretty strange, I do remember checking for such things in the past and we were behaving the same with xterm, however maybe not all 'clearing' escapes should do that with 'only bg' template? Underlines, etc are also not cleared afaics.

```
echo -ne "\e[38;2;0;255;255m\e[48;2;0;0;0m\a\e[K";sleep 10
```

I feel like they preserve foreground color? Not sure about flags, they unlikely preserve that, and I do remember testeing for such things i nthe past, however I don't remember If we actually verified foreground color...

**Comment by chrisduerr** [comment]
Seems like this is all you need actually:

```
echo -ne "\e[32m\e[41m\e[K" && sleep infinity
```

So truecolor does not matter here.

Cursor motion also doesn't seem to reset the cell state, so it might just be all about the line reset call (you might need to change the last escape to `\e[H` when not using XTerm with backspace):

```
reset && echo -ne "\e[32m\e[41m\e[K\e[10H\e[2H"; sleep infinity
```

If the background color is the default, the cursor color will not be changed though:

```
echo -ne "\e[32m\e[K"; sleep infinity
```

When explicitly setting a background color, but the background color matches the terminal's default background, the cursor's color is also changed (this is for XTerm's white colorscheme):

```
echo -ne "\e[107m\e[32m\e[K"; sleep infinit
```

So unless you have other suggestions @kchibisov, I think this should be all the information we need? It seems like the logic's basically to reset the foreground color of the cell immediately (without writing to it), whenever the background color is not the default.

Though I'm not entirely sure about *why* it is done this way. It seems like all the terminals that do reset it conditionally do not **always** reset it, which seems strange to me.

**Comment by kchibisov** [comment]
>So unless you have other suggestions @kchibisov, I think this should be all the information we need? It seems like the logic's basically to reset the foreground color of the cell immediately (without writing to it), whenever the background color is not the default.

>Though I'm not entirely sure about why it is done this way. It seems like all the terminals that do reset it conditionally do not always reset it, which seems strange to me. Perhaps it's about not changing the selection color unless necessary?

Yeah, that's clear what I've mentioned on IRC, it was also mentioned that not all terminals do that thing so it seems like 'someone started weird impl and everyone decided to follow it' and clients started to rely on it, since it's purely insanity to rely on terminal that resets things when there's no background color(aka default), but that doesn't reset when you have some...

We may add such insane logic, but knowing that some terminals(urxvt,kitty) do the same thing as we are it's also questionable.

**Comment by chrisduerr** [comment]
> We may add such insane logic, but knowing that some terminals(urxvt,kitty) do the same thing as we are it's also questionable.

Yeah, it might be better to just let applications handle this difference. Alternatively it would make some sense to always override the foreground if possible. But this in-between seems rather strange.

Another interesting thing is that even when the cursor isn't changed in color (`echo -ne "\e[32m\e[K"; sleep infinity`), when selecting things the cells still invert to the foreground color. So the information is apparently **always** updated, just not the cursor.

**Comment by kevinhwang91** [comment]
I use onedark theme in neovim encounter this issue, reproduce steps:
1. `alacritty --config-file reproduce.yml`
2. `echo -ne '\e[38;2;0;0;0m\e[48;2;229;192;123m\a\e[K'; sleep infinity`
3. `echo -ne '\e[38;2;0;0;0m\e[48;2;209;154;102m\a\e[K'; sleep infinity`


`cat reproduce.yml`
```yaml
colors:
  primary:
    background: '#202326'
    foreground: '#abb2bf'
  cursor:
   text: '#000000'
   cursor: '#528bff'
```
![image](https://user-images.githubusercontent.com/17562139/102120083-fdcefe00-3e7c-11eb-9732-df4aa7f7f440.png)

-----------------------------------------------------------------------------------------------------------------------------------

![image](https://user-images.githubusercontent.com/17562139/102120130-0d4e4700-3e7d-11eb-8076-3b1a26642933.png)

https://github.com/alacritty/alacritty/pull/4311 don't work for my issue.

**Comment by kevinhwang91** [comment]
Would this issue be fixed?

**Comment by PranavBhattarai** [comment]
![Screenshot from 2021-05-31 12-01-13](https://user-images.githubusercontent.com/32763112/120148447-f96a8780-c207-11eb-82cf-e711a6694059.png)
I was having the same issue in Alacritty + Fish.
After my research I found that the theme which I installed using the command `npx alacritty-themes` have this inside alacrity.yml:
```
cursor:
    text: "#FF0000"
    cursor: "#FF0000"
```
Both text and cursor color were the same which made words completely unreadable if there is a cursor on any word. 
I altered them using @kevinhwang91 text and cursor color sets. Now I could see fairly.
